File: lab3/src/main/scala/ru/ifmo/ctddev/komarov/net/lab3/repl/REPL.scala
package ru.ifmo.ctddev.komarov.net.lab3.repl

import java.io.{BufferedReader, InputStreamReader}

trait REPL[T] {

  val parser: String => Option[T]
  val handler: Option[T] => Unit

  def main(args: Array[String]) {
    val r = new BufferedReader(new InputStreamReader(System.`in`))
    while (true) {
      print(">>> ")
      handler(parser(r.readLine()))
    }
  }
}

File: lab3/src/lab3/repl/REPL.java
package lab3.repl;

import lab3.main.World;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class REPL implements Runnable {
    private final World world;

    public REPL(World world) {
        this.world = world;
    }

    @Override
    public void run() {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            try {
                String s = br.readLine();
                if (s == null) {
                    System.out.println("Bye!");
                    System.exit(0);
                }
                Request rq = RequestParser.parse(s);
                rq.apply(world);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}
